layerscape: align layerscape source code to layerscape lsdk

yangbolu1991 opened the pull request:
This patch-set is to update LEDE layerscape source code with NXP Layerscape LSDK.
https://lsdk.github.io
https://github.com/qoriq-open-source

Attached verification logs for current 5 layerscape platforms.

[layerscape-log.zip](https://github.com/lede-project/source/files/1152515/layerscape-log.zip)

yangbolu1991 commented:
Any comments on this patch-set?
:)

nbd168 commented on package/boot/uboot-layerscape-armv8_32b/Makefile:
```diff
@@ -8,18 +8,18 @@
-PKG_SOURCE_URL:=https://github.com/fsl-jyt/uboot-ls-32b.git
-PKG_SOURCE_VERSION:=6476195551ea83193afd06d0c441483c04729c8d
+PKG_MIRROR_HASH:=e8a3294cae1aa0b47fbe159f75051060ab2a81e0a52f7609f5306bd65428f636
+PKG_SOURCE_URL:=git@github.com:yangbolu1991/u-boot-lede.git
```
Please fix the source URL, it should refer to a public one that does not require a github account.

nbd168 replied:
What I mean is:
`https://github.com/yangbolu1991/u-boot-lede.git` instead of `git@github.com:yangbolu1991/u-boot-lede.git`.
The first one allows public download, the second one requires a ssh key for github access.

yangbolu1991 replied:
Oh... Sorry for my misunderstanding. I will update the link in next version.

hauke commented on package/boot/uboot-layerscape-armv8_32b/Makefile:
```diff
@@ -0,0 +1,51 @@
+	$(INSTALL_DIR) $(STAGING_DIR_IMAGE)
+	$(CP) $(PKG_BUILD_DIR)/ls1043ardb-uboot.bin $(STAGING_DIR_IMAGE)/ls1043ardb-armv8_32b-uboot.bin
+	$(CP) $(PKG_BUILD_DIR)/ls1046ardb-uboot.bin $(STAGING_DIR_IMAGE)/ls1046ardb-armv8_32b-uboot.bin
+	$(CP) $(PKG_BUILD_DIR)/ls1012ardb-uboot.bin $(STAGING_DIR_IMAGE)/ls1012ardb-armv8_32b-uboot.bin
```
Why don't you build these binaries from source code inside of LEDE?
For GPL compliance you anyway have to provide the source code.

yangbolu1991 replied:
We always used 64b u-boot image to support both 32b and 64b kernel and rootfs. It's convenient.
So this uboot-layerscape-armv8_32b here used same image/code with uboot-layerscape-armv8_64b.
The reason we used image is that LEDE could only provide 32b toolchain for 32b target.

yangbolu1991 replied:
Any suggestion here? Or should I add a note to explain in Makefile?

hauke replied:
this is ok with me

hauke commented on package/firmware/fman-ucode/Makefile:
```diff
@@ -9,13 +9,13 @@ include $(TOPDIR)/rules.mk
+PKG_SOURCE_VERSION:=c275e91392e2adab1ed22f3867b8269ca3c54014
 
 PKG_BUILD_DIR=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
 
```
The License tag is wrong:
PKG_LICENSE:=GPL-2.0 GPL-2.0+This is not the GPL: https://github.com/qoriq-open-source/fm-ucode/blob/integration/Freescale-Binary-EULA

yangbolu1991 replied:
Oh, you're right... These're binaries developed by Freescale/NXP.
Let me remove the error PKG_LICENSE.

May I know whether PKG_LICENSE is required in Makefile? Could I just ignore it.
fman-ucode is neither GPL or BSD, it's just Freescale-Binary-EULA.

hauke replied:
This looks ok to me now

hauke commented on package/firmware/rcw/Makefile:
```diff
@@ -9,13 +9,13 @@ include $(TOPDIR)/rules.mk
+PKG_SOURCE_VERSION:=6eed8fdfc07920178b70fba199fa20fb42adb20b
 
 PKG_BUILD_DIR=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
 
```
This package is licensed under some BSD license and not GPL:
https://github.com/qoriq-open-source/rcw/blob/integration/LICENSE

yangbolu1991 replied:
I will correct it in next version. :)

hauke commented on package/firmware/rcw/Makefile:
```diff
@@ -78,6 +78,7 @@ define Build/Configure
 
 define Build/Compile
+	$(MAKE) -C $(PKG_BUILD_DIR)
 endef
```
you can remove the "define Build/Compile" block completely, this is more or less what "define Build/Compile/Default" does.

yangbolu1991 replied:
Thanks. Will remove this in next version.

hauke commented on target/linux/layerscape/armv8_32b/target.mk:
```diff
@@ -0,0 +1,15 @@
+
+ARCH:=arm
+BOARDNAME:=ARMv8 32-bit based boards
+CPU_TYPE:=cortex-a9
```
What type of CPUs are supported by this target?
Is it a cortex A9 or some other CPU like a cortex A53?

yangbolu1991 replied:
Layerscape CPU is cortex A53 or cortex A72 now, but we want it to run on 32-bit mode which is compatible with ARMv7.
So we set CPU_TYPE as cortex-a9 to select 32b toolchain.

yangbolu1991 replied:
@chunkeey 
Sorry for my late response. I just come back to office today from holiday.
I didn't insist A9 ... In current layerscape platforms, some are with A53 cpus and some are with A72 cpus. More details,
http://www.nxp.com/products/microcontrollers-and-processors/arm-processors/qoriq-layerscape-arm-processors:QORIQ-ARM

I don't know which CPU_FLAGS is proper for *all* of them and doesn't introduce any potential issue, but at least platforms worked fine with current A9.
You suggested to use A7 or A15, which one exactly should I use in your view?
Should I also add CPU_SUBTYPE:=neon-vfpv4 ?
I really would like to fix it with proper value.

Regarding to BOARDNAME, I think it seems proper because they are indeed ARMv8 32-bit.

@hauke @nbd168 @fsl-jyt 
Do you have any suggestion on CPU_TYPE?
Thanks a lot :)

hauke replied:
As the layerscape 32 biot target is not so popular I would suggest choosing something which is already used by some other target, see here: https://downloads.lede-project.org/snapshots/packages/

I would suggest this:
CPU_TYPE:=cortex-a15
CPU_SUBTYPE:=neon-vfpv4 
This is probably a better fit than the A9

yangbolu1991 replied:
Thanks a lot for @chunkeey @hauke  suggestion! 
It's ok for me. I will fix this in next version.

hauke commented on target/linux/layerscape/armv8_64b/config-default:
```diff
@@ -0,0 +1,4667 @@
+#
```
Please run "make kernel_oldconfig CONFIG_TARGET=subtarget" when this subtarget is selected to reduce the size of this configuration file.

yangbolu1991 replied:
Thanks a lot for letting me know this method.
No problem.

hauke commented on target/linux/layerscape/armv8_32b/config-default:
```diff
@@ -0,0 +1,6337 @@
+#
```
Please run "make kernel_oldconfig CONFIG_TARGET=subtarget" when this subtarget is selected to reduce the size of this configuration file.

yangbolu1991 replied:
Thanks a lot for letting me know this method.
No problem.

hauke commented on target/linux/layerscape/armv8_64b/target.mk:
```diff
@@ -0,0 +1,15 @@
+
+ARCH:=aarch64
+BOARDNAME:=ARMv8 64-bit based boards
+CFLAGS:=-Os -pipe -fno-caller-saves
```
Is the CFLAGS setting really needed? LEDE should use similar compiler options by default.

yangbolu1991 replied:
I just tried to compile without that. There was no problem. Let me remove it in next version. Thanks.

fsl-jyt replied:
Hi @yangbolu1991 ,
Please double check the libnl-tiny-0.1
Without that CFLAGS, it will arise compile error, please have a look: 
https://github.com/lede-project/source/pull/344

nbd168 replied:
What kind of compile error? I don't see an error log anywhere in there...

yangbolu1991 replied:
@fsl-jyt Thanks for providing the information. 
I just double checked the compiling(make distclean and make, for NXP target) and no problem. Maybe the old issue had been fixed.

So I think there should be no problem to remove it.

hauke replied:
yes please remove this

hauke commented on target/linux/layerscape/armv8_64b/target.mk:
```diff
@@ -0,0 +1,15 @@
+# See /LICENSE for more information.
+#
+
+ARCH:=aarch64
```
If the SoC supported by this target all have an cortex A53 CPU, please add:
CPU_TYPE:=cortex-a53

yangbolu1991 replied:
Actually, ls1012a/ls1088a/ls1043a is cortex A53 cpu while ls1046a/ls2088a is cortex A57 cpu.

hauke replied:
I am ok with this, then you can share the packages with the armvirt target.

hauke commented on target/linux/layerscape/patches-4.4/0001-arch-Introduce-post-init-read-only-memory.patch:
```diff
@@ -0,0 +1,114 @@
+From 7ece66e296972744ec471d521b8ca07990e2494f Mon Sep 17 00:00:00 2001
```
This commit and probably many more are already in kernel 4.9, could you please update this target to kernel 4.9.
The next LEDE release will be based on kernel 4.9 anyway so you have to update this target when the layerscape target should be included in the next release.

yangbolu1991 replied:
NXP LSDK will release kernel v4.9 on September 2017. We will update our targets to v4.9 in LEDE on September.
But currently customers still use v4.4, this patch-set is still needed.

May I know whether LEDE will still maintain v4.4 after v4.9 is released?

hauke replied:
I do not like all these patches being added you are ending up with 835 patches on top of kernel 4.4. Please do the migration to kernel 4.9 then you can probably get rid of many patches.

yangbolu1991 replied:
@hauke , 
Almost all these patches are for layerscape IP modules. We are also working on kernel 4.9 now. Even we migrate to kernel 4.9, I'm afraid the number of patches would not shrunk so much. 

The large number patches mainly comes from dts (155 patches), mc-bus (139 patches), sec (138 patches), qspi (80 patches). What I can do is to merge each IP module patches into one. We can provide only one patch for each module/component. Then we will only have dozens of patches finally.

It's also easy for maintaining, becasue we have each module as a feature branch when backport patches.

Do you think is it ok? Let me generate a new version for your reveiwing.
Thank you very much.

yangbolu1991 commented:
@hauke @nbd168 @fsl-jyt 
I updated the patch-set with below changes to address your comments.
Please help to review. Thanks a lot :)

layerscape: take over maintainership
Changes for v2:
	- None

layerscape: rename subtargets and update makefile files
Changes for v2:
	- None

layerscape: update packages with LSDK git trees
Changes for v2:
	- Fixed u-boot-layerscape-armv8_32b url with public url.
	- Added note to explain why used u-boot images for 32b targets, and removed wrong license info.
	- Removed wrong license info of fman_ucode.
	- Corrected rcw license info with BSD.
	- Removed useless Build/Compile definition for rcw.
	
layerscape: update memory layout according to LSDK
Changes for v2:
	- None

layerscape: update linux to LSDK linux
Changes for v2:
	- Reduced size of kernel config for 32b/64b targets.
	- Removed useless CFLAGS for 64b targets.

yangbolu1991 commented:
Any comments? 
Thanks a lot.

yangbolu1991 commented:
Just rebased this patch-set.

yangbolu1991 commented:
Could we consider to merge this pull request if there is no other change request?
This patch-set is mainly to update all packages code and kernel patches to Layerscape SDK.
Regarding to CPU_TYPE discussion, we can still continue or start a email thread after this PR was closed, and I can send a seperate patch for that if need.

Thanks a lot :)

yangbolu1991 commented:
I have updated the patch-set with v3 version.
See below changes for v3,

layerscape: take over maintainership
 Changes for v3:
 - None

layerscape: rename subtargets and update makefile files
 Changes for v3:
 - None

layerscape: update packages with LSDK git trees
 Changes for v3:
 - None

layerscape: update memory layout according to LSDK
 Changes for v3:
 - None

layerscape: update linux to LSDK linux
 Changes for v3:
 - fixed CPU_TYPE issue for 32-bit.
 - merged each IP module kernel patches into one.

yangbolu1991 commented:
@hauke 
v3 patch-set fixed the CPU_TPYE problem, and merge each IP module patches into one. Now there are only 36 patches.
Do you think is it ok?

yangbolu1991 commented:
Hi @hauke ,
Sorry for disturbing.  Did you have a chance to have a look at this?
Thanks.

yangbolu1991 commented:
@hauke @chunkeey @nbd168 @fsl-jyt 
Hi all,

Rebased this patch-set again since Hauke's below patch introduced conflicts. Actually this changes had already in this patch-set as we discussed.
a5d26d7 layerscape: do not add custom CFLAGS.

Could we speed up the reviewing? I think all current comments had been addressed, right?
Thanks a lot.

yangbolu1991 commented:
I splited patch "layerscape: update linux to LSDK linux" into 5 patches.
I think it will be easier for you to review.

See https://github.com/lede-project/source/pull/1327
